### PR TITLE
Create Projectoid.js

### PR DESCRIPTION
**Pull request by RoyalDevendra**
Introducing the **Projectoid** Library for Telegram Bots! This versatile JavaScript BB library empowers bot developers to seamlessly integrate their Telegram bots with Projectoid, unlocking a world of possibilities. With a focus on simplicity and functionality, this library offers a suite of features to enhance your bot's capabilities.

_Key Features:_

- **API Calls:** Execute API calls to Projectoid effortlessly. Make GET and POST requests with ease using the library's intuitive methods.

- **Access Token Management:** Simplify access token management by saving and retrieving tokens seamlessly, ensuring secure and authenticated interactions with Projectoid.
- **Broadcast Messages:** Streamline message broadcasting to multiple chats using various methods, including forwarding and copying messages.
- **Membership Checks:** Verify user membership in chats, making it easy to control access and permissions.

_With this library, you can harness the full potential of Projectoid in your Telegram bot, enhancing user engagement and expanding the functionality of your chatbot applications._

### CODE REVIEW

File: Projectoid.js [hunk 1]
```diff
@@ -0,0 +1,313 @@
+//The Global Variables
+let libPrefix = "projectoid";
```
Comment (bots-business) [inline Projectoid.js:2]: use const her not let

File: Projectoid.js [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+let API_URL = "https://api.projectoid.site/v1";
+
+// Function to convert JSON to URL query string
+function jTQS(jsonData) {
```
Comment (bots-business) [inline Projectoid.js:6]: bad fun name. 

use convertJsontToUrlQuery or same

File: Projectoid.js [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+    .join("&");
+  return queryString;
+}
+
+/**
+ * apiCall() - Makes an API call to Projectoid.
+ *
+ * @since 1.0.0
+ *
+ * @param {object} options - The API call options.
```
Comment (bots-business) [inline Projectoid.js:20]: need example for options here

File: Projectoid.js [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+ */
+function apiCall(options) {
+  if (!options) {
+    throw libPrefix + ": apiCall: options not found";
```
Comment (bots-business) [inline Projectoid.js:24]: options not passed ?

File: Projectoid.js [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+  }
+  let apiData = {
+    url: url,
+    headers: {
+      ...(options.method.toLowerCase() === "get"
+        ? {}
+        : { "Content-Type": "application/json" }),
+    },
+    body: options.body,
+    folow_redirects: true,
```
Comment (bots-business) [inline Projectoid.js:38]: redirect is really needed?

File: Projectoid.js [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+ * @param {string} command - The command to execute on success.
+ */
+function addChat(chatid, access_token, command) {
+  if (isNaN(parseInt(chatid))) {
+    throw libPrefix + ": addChat: incorrect chat id";
+  }
+
+  if (!access_token) {
+    var access_token = Bot.getProperty("Projectoid_AccessToken");
+    if (!access_token) {
```
Comment (bots-business) [inline Projectoid.js:86]: your already have if (!access_token)  before this. 

this line not needed

File: Projectoid.js [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+
+  let requestData = {
+    method: "post",
+    path: "telegram/botpanel/adduser.php",
+    body: {
+      bot_id: bot.token.split(":")[0],
+      user_id: chatid,
+      access_token,
+    },
+    onSuccess: !command ? null : command,
```
Comment (bots-business) [inline Projectoid.js:99]: use `command ? command : null`

File: Projectoid.js [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+  if (options.method == "forwardMessage") {
+    initiateForwardBroadcast(options);
+    return;
+  }
+  if (options.method == "copyMessage") {
+    initiateCopyBroadcast(options);
+    return;
+  }
+
+  let text = !options.text ? null : options.text;
```
Comment (bots-business) [inline Projectoid.js:128]: const  text = options.text - it is enough ?

File: Projectoid.js [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+  let type = !options.type ? null : options.type;
+  let file_id = !options.file_id ? null : options.file_id;
+  let caption = !options.caption ? null : options.caption;
+  let parseMode = !options.parseMode ? null : options.parseMode;
```
Comment (bots-business) [inline Projectoid.js:132]: { text, file_id, parseMode ... } = options

it can be more simple here

File: Projectoid.js [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+  let disableWebPreview = !options.disableWebPreview
+    ? null
+    : options.disableWebPreview;
+  let protectContent = !options.protectContent ? false : options.protectContent;
+  let webhookUrl = !options.webhookUrl ? null : options.webhookUrl;
+  let cmd = !options.command ? null : options.command;
+
+  if (!options.access_token) {
+    var access_token = Bot.getProperty("Projectoid_AccessToken");
```
Comment (bots-business) [inline Projectoid.js:141]: use:

const token = options.access_token || Bot.getProperty("Projectoid_AccessToken");
if(!token){
   throw ...
}

File: Projectoid.js [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+  let message_id = options.message_id;
+
+  if (!from_chat_id || !message_id) {    
+    throw libPrefix + ": initiateForwardBroadcast: chat id or message id was not found";
+  }
+  let protectContent = !options.protectContent ? false : options.protectContent;
+  let webhookUrl = !options.webhookUrl ? null : options.webhookUrl;
+  let cmd = !options.command ? null : options.command;
+
+  if (!options.access_token) {
```
Comment (bots-business) [inline Projectoid.js:184]: you need to move this code to function because you use it again and again

File: Projectoid.js [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+  }
+}
+
+// Function called when an API request results in an error
+function onApiError() {
+  throw libPrefix + ": " + content + "\nGet Help at @ProjectoidChat";
+}
+
+publish({
+  apiCall: apiCall,
```
Comment (bots-business) [inline Projectoid.js:303]: descriptions is needed below all methods here. 

hard to understand why we need this methods

File: Projectoid.js [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+      userId: options.userId,
+      chatIds: options.chats,
+    },
+    onSuccess: options.command,
+  };
+  apiCall(requestData);
+}
+
+// Function called when an API answer is received
+function onApiAnswer() {
```
Comment (bots-business) [inline Projectoid.js:284]: use "_" for all private methods:

_onApiAnswer()

**Review by bots-business — CHANGES_REQUESTED** [review]
Much more needed changes

File: Projectoid.js (hunk 4 above)
Comment (RoyalDevendra) [reply Projectoid.js:24]: what do you means options are not passed?

This function can be used directly in bot too, that is why we need to check if the options are passed or not.

File: Projectoid.js (hunk 5 above)
Comment (RoyalDevendra) [reply Projectoid.js:38]: yes it is needed.

i don't know why, but sometimes in bb it creates a issue if don't use it.

File: Projectoid.js (hunk 8 above)
Comment (RoyalDevendra) [reply Projectoid.js:128]: no (sometimes) - to prevent the errors saying `undefined "options.something"`

**Comment by RoyalDevendra** [comment]
Easily integrate Projectoid's powerful features into your Telegram bot with this BB library. Perform API calls to Projectoid and effortlessly broadcast messages. Additionally, efficiently conduct User Membership checks. Simplify your bot development and enhance the chatbot experience with Projectoid using this convenient library.

**Comment by RoyalDevendra** [comment]
Creating Another PR with some fixes

**Pull request closed without merging**